snippets/scripts/jira-components-to-compass-components/migrateJiraComponentsToCompassComponents.py

```diff
@@ -0,0 +1,329 @@
+        issues = get_related_issues_for_component(PROJECT_KEY, component)
+        component_id = component['id']
+
+        # filter out issues with 0 components
+        issues = [issue for issue in issues if len(issue['fields']['components']) > 0]
+
+        if len(issues) > 0:
+            jira_component_to_issues_map[component_id] = {
+                'component': component,
+                'issues': issues
+            }
+
+    # print for each component how many issues are found
+    for component_id, component in jira_component_to_issues_map.items():
+        if len(component['issues']) > 0:
+            print(f"Jira Component {component['component']['name']} has {len(component['issues'])} issues")
+
+    print("\n")
+
+    # create a map of jira component id to None
+    jira_components_to_compass_components_map = {}
+
+    # for each jira component, see if there is a compass component that matches the name
+    # create a map of jira component id to compass component
+    for jira_component in jira_components:
+        compass_component = find_compass_component_in_jira(jira_component['name'])
+        if compass_component is not None:
+            jira_components_to_compass_components_map[jira_component['id']] = compass_component
+
+    if len(jira_components_to_compass_components_map) > 0:
+        print("We found following Compass components with same name as the Jira components:")
+    for jira_component in jira_components:
+        if jira_component['id'] in jira_components_to_compass_components_map:
+            print(f"{jira_component['name']}")
+
+    print("\n")
+    # find out components that needs to be created in Compass
+    jira_components_to_create_in_compass = [jira_component for jira_component in jira_components
+                                            if jira_component['id'] not in
+                                            jira_components_to_compass_components_map]
+
+    if len(jira_components_to_create_in_compass) > 0:
+        print(f"Corresponding Compass components for following Jira components will be created:")
+        for jira_component in jira_components_to_create_in_compass:
+            print(f"{jira_component['name']}")
+        print(f"To view newly created Compass components, visit: "
+              f"https://{DOMAIN_NAME}/compass/components?label=migrated-from-jira-component")
+
+    print("\n")
+
+    for jira_component in jira_components_to_create_in_compass:
+        compass_component = create_compass_component(cloud_id, jira_component)
+        if compass_component['success'] is False:
+            print(f"Failed to create compass component for Jira Component"
+                  f" with id {jira_component['id']} and name {jira_component['name']}")
+            print(f"Error: {compass_component['error']}")
+
+    # sleep a bit for the components to sync in Jira and retrieve them as Jira Components again
+    time.sleep(len(jira_components_to_create_in_compass) // 10 + 1)
```

Comment:
Hmmmm, this may be a bit troublesome 😬 I had a run where the component was created in Compass, by the sync delay was fairly long and the component did not get synced to Jira until (somewhat long) after the script finished running, so the issue was never updated. Re-running the script after the sync finished managed to migrate the issues.

Maybe run a polling loop that checks whether the component(s) have been created instead? 🙃 

Or add a disclaimer that if issues weren't migrated, the user should re-run the script.

Reply:
ya I saw that a couple of times as well. Since the script can be re-run multiple times, I guess we can just update the readme and call it day 🤷🏾